File: Luogu/P1443/P1443.py
from collections import deque
import numpy as npy

class Node:
    def __init__(self, x, y, step):
        self.x = x
        self.y = y
        self.step = step

dx = [2, 1, -1, -2, -2, -1, 1, 2]
dy = [1, 2, 2, 1, -1, -2, -2, -1]

n, m, sx, sy = map (int, input().split())
sx -= 1
sy -= 1
ans = npy.full ((405, 405), -1)

def bfs ():
    q = deque()
    q.append (Node (sx, sy, 0))
    ans[sx, sy] = 0
    while q:
        cur = q.popleft()

        for i in range (8):
            nx = cur.x + dx[i]
            ny = cur.y + dy[i]
            if 0 <= nx < n and 0 <= ny < m and ans[nx, ny] == -1:
                ans [nx, ny] = cur.step + 1
                q.append(Node (nx, ny, cur.step + 1))
bfs ()

for i in range(n):
    for j in range(m):
        print(f"{ans[i, j]:-5d}", end = "")
    print()

File: Luogu/P1443/P1443.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <queue>
using namespace std;

struct Node 
{
    int x, y, step;
};

int dx[] = {2, 1, -1, -2, -2, -1, 1, 2}; 
int dy[] = {1, 2, 2, 1, -1, -2, -2, -1}; 
int ans[405][405];
int n, m, sx, sy;

void bfs() 
{
    queue<Node> q;
    q.push({sx, sy, 0});
    ans[sx][sy] = 0;

    while (!q.empty()) 
    {
        Node cur = q.front();
        q.pop();

        for (int i = 0; i < 8; i++) 
        {
            int nx = cur.x + dx[i];
            int ny = cur.y + dy[i];
            if (nx >= 0 && nx < n && ny >= 0 && ny < m && ans[nx][ny] == -1) 
            {
                ans[nx][ny] = cur.step + 1;
                q.push({nx, ny, cur.step + 1});
            }
        }
    }
}

int main() 
{
    memset(ans, -1, sizeof(ans)); 
    cin >> n >> m >> sx >> sy;
    sx--; sy--; 
    bfs(); 
    for (int i = 0; i < n; i++) 
    {
        for (int j = 0; j < m; j++) 
        {
            printf("%-5d", ans[i][j]); 
        }
        printf("\n");
    }
    return 0;
}